reana_job_controller/utils.py

```diff
@@ -69,6 +73,118 @@ def initialize_krb5_token(workflow_uuid):
+    :type replacements: dict
+    :param skip_initial_space: ignore whitespace immediately following the delimiter
+    :type skip_initial_space: bool
+    :param skip_trailing_space: ignore whitespace at the end of each csv row
+    :type skip_trailing_space: bool
+    """
+    if skip_trailing_space:
+        input_csv = "\n".join((line.strip() for line in input_csv.splitlines()))
+
+    replacements = replacements or {}
+    with StringIO(input_csv) as csv_input:
+        csv_reader = csv.DictReader(
+            csv_input,
+            fieldnames=fieldnames,
+            delimiter=delimiter,
+            skipinitialspace=skip_initial_space,
+        )
+        for row in csv_reader:
+            yield {
+                key: value if value not in replacements.keys() else replacements[value]
+                for key, value in row.items()
+            }
+
+
+def motley_cue_auth_strategy_factory(hostname):
+    """
+    Paramiko auth strategy factory that provides oauth based ssh token authentication.
+
+    This auth strategy has been developed against the motley cue implementation of
+    oauth based ssh token authentication on the server side.
+
+    :param hostname: hostname of the ssh node
+    :type hostname: str
+    """
+    # Using a factory to avoid a general dependency on libmytoken, paramiko and pyjwt
+    from libmytoken import get_access_token_from_jwt_mytoken
+    from paramiko.auth_strategy import AuthSource
+    from time import time
+    import jwt
+    import requests
+
+    class MotleyCueTokenAuth(AuthSource):
+        def __init__(self):
+            self._access_token = None
+            self._access_token_expires_on = 0
+            self.hostname = hostname
+            self.username = self._get_deployed_username()
+            super().__init__(username=self.username)
+
+        @property
+        def access_token(self):
+            if not (self._access_token and self._is_access_token_valid()):
+                self._refresh_access_token()
+            return self._access_token
+
+        def authenticate(self, transport):
+            return transport.auth_interactive(
+                username=self.username, handler=self.motley_cue_auth_handler
+            )
+
+        def _is_access_token_valid(self):
+            return (
+                self._access_token_expires_on - time() > 100
+            )  # token should be at least valid for 100 s
+
+        def _get_access_token_expiry_date(self):
+            decoded_token = jwt.decode(
+                self._access_token,
+                options={"verify_signature": False, "verify_aud": False},
+            )
+            return decoded_token["exp"]
+
+        def _get_deployed_username(self):
+            headers = {"Authorization": f"Bearer {self.access_token}"}
+            req = requests.get(
+                f"https://{self.hostname}/user/deploy", headers=headers, verify=True
+            )
+            return req.json()["credentials"]["ssh_user"]
```

Comment:
I think here you should probably check whether the server returned an error status code (4xx, 5xx) or not

Reply:
Thanks, has been fixed in the latest update by using `req.raise_for_status()` before the return statement.